doc/python: explain bulk package set updates (#127427)
NixOS/nixpkgs: doc/languages-frameworks/python.section.md

  If necessary, `pname` has to be given a different value within `fetchPypi`.
* Attribute names in `python-packages.nix` should be sorted alphanumerically to
  avoid merge conflicts and ease locating attributes.

## Package set maintenance

> The whole python package set has alot of packages that does not see regular

Reviewer: ```suggestion
The whole Python package set has a lot of packages that does not see regular
```

In other section "Python" starts with a capital P.

> updates, because they either are a very fragile component in the python

Reviewer: ```suggestion
updates, because they either are a very fragile component in the Python
```

ecosystem, like for example the `hypothesis` package, or packages that have
no maintainer, so maintenance falls back to the package set maintainers.

### Updating packages in bulk

> There is a tool to update alot of python libraries in bulk, it exists at

Reviewer: ```suggestion
There is a tool to update a lot of Python libraries in bulk, it exists at
```

> `./maintainers/scripts/update-python-libraries` with this repository.

Reviewer 2: ```suggestion
`maintainers/scripts/update-python-libraries` with this repository.
```

It can quickly update minor or major versions for all packages selected

> and create update commits, and supports the `fetchPypi`, `fetchUrl` and

Reviewer 3: ```suggestion
and create update commits, and supports the `fetchPypi`, `fetchurl` and
```

> `fetchFromGitHub` fetchers.

Author: ```suggestion
`fetchFromGitHub` fetchers. When updating lots of packages that are hosted on GitHub, exporting a `GITHUB_API_TOKEN` is recommended.
```

These bulk updates take place in the `python-unstable` branch which should

> normally be merged into `staging-next` after a stabilization period.

Reviewer 3: staging would be more ideal. Staging-next was chosen this time around, just because a staging-next was started a few days prior

An exemplary call to update all python libraries between minor versions
would be:

> ```

Reviewer 2: ````suggestion
```ShellSession
````

> ./maintainers/scripts/update-python-libraries --target minor --commit --use-pkgs-prefix pkgs/development/python-modules/**/default.nix

Reviewer 2: ```suggestion
$ ./maintainers/scripts/update-python-libraries --target minor --commit --use-pkgs-prefix pkgs/development/python-modules/**/default.nix
```

> ```

Reviewer 3: should add that after doing the mass update, you will still need to stabilize. This was was hinted earlier, but should probably make it explicit that the builk update is a very big hammer, and you likely broke many packages.

  If necessary, `pname` has to be given a different value within `fetchPypi`.
* Attribute names in `python-packages.nix` should be sorted alphanumerically to
  avoid merge conflicts and ease locating attributes.

## Package set maintenance

The whole Python package set has a lot of packages that do not see regular
updates, because they either are a very fragile component in the Python
ecosystem, like for example the `hypothesis` package, or packages that have
no maintainer, so maintenance falls back to the package set maintainers.

### Updating packages in bulk

There is a tool to update alot of python libraries in bulk, it exists at
`maintainers/scripts/update-python-libraries` with this repository.

It can quickly update minor or major versions for all packages selected
and create update commits, and supports the `fetchPypi`, `fetchurl` and
`fetchFromGitHub` fetchers. When updating lots of packages that are

> hosted on GitHub, exporting a `GITHUB_API_TOKEN` is recommended.

Reviewer 2: ```suggestion
hosted on GitHub, exporting a `GITHUB_API_TOKEN` is highly recommended.
```

Author: Applied.

Updating packages in bulk leads to lots of breakages, which is why a
stabilization period on the `python-unstable` branch is required.

Once the branch is sufficiently stable it should normally be merged
into the `staging` branch.

An exemplary call to update all python libraries between minor versions
would be:

> ```shell

Reviewer 2: ````suggestion
```ShellSession
````

Author: Please explain the difference.

Reviewer 2: shell is the real name of bash and ShellSession is for shell examples with $ and maybe output. See https://github.com/github/linguist/blob/fec30b07654881001603ce05cc772b56d8d87060/lib/linguist/languages.yml#L5594
